Setup Wizard

HazAT opened the pull request:
The idea behind the Sentry Wizard is that users no longer have to do any manual copy/pasting or reading docs to get any SDK up and running.
Sentry Wizard is a command line tool which fetches all necessary info needed for the config of a SDK.
https://github.com/getsentry/sentry-wizard

![wizard mov](https://user-images.githubusercontent.com/363802/32745484-2f9fef2c-c8b3-11e7-92c4-9286921fbd1e.gif)

So this PR adds the setup-wizard endpoint.
Here is a flow diagram of how it works:

![untitled diagram 3](https://user-images.githubusercontent.com/363802/32718704-318ec164-c85e-11e7-9508-b0a19bcdbe0e.png)


- [x] Merged https://github.com/getsentry/sentry/pull/6541

getsentry-danger commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

HazAT commented:
@getsentry/platform this is probably the most crucial part:
https://github.com/getsentry/sentry/pull/6523/files#diff-71b608da8ab4ef6157551bd8445bd9f5R60
I needed this because the redirect back to the wizard from the login page was broken.

cc @ehfeng How would you prefer that we add "analytics" to this.
Should we track something when someone opens the wizard page?
OR
Should the wizard send "analytics" data to an endpoint when the user finished it?

HazAT commented:
@mitsuhiko I am now doing internal API call instead of fetching it from the frontend.

@getsentry/platform I've added a finished state to the loading indicator to have a fancy checkmark.

mattrobenolt commented:
Can we discuss if we should merge this feature first? Not a fan of Hack Week immeditely turning into “now let’s merge it.”

HazAT commented:
@mattrobenolt Absolutely, what are your concerns/what do you want me to explain so we can get this landed?
I hope the reason why I made this was clear and it's kind of a "blocker" for the new Cordova SDK since I want to use this feature there.
If there are reasons not to merge it we can discuss it but not just "not merge because it was a Hack week project".

mitsuhiko commented:
So some notes from me on this:

* I'm generally happy with the concept and I do not mind the implementation
* I agree that we want this for the onboading flow of new clients
* I do think someone else should sign this off with regards to the security aspects of it

Additionally since this will not work with on-prem for quite some time I do think we should launch this on the server first and not depend on the clients for an extended period of time until this made it into stable sentry. Likewise we might want to add an endpoint to be able to probe if the server supports this feature to provide a fallback flow for the actual client based wizard code (eg: to give the user a "server does not support this feature, i'm sorry, here is what you can do instead" message in the terminal).

HazAT commented:
@mitsuhiko Why should it not work for on-prem customers?
The wizard takes a param `-u` or `--url` where you can define the URL of your Sentry installation, so this should work.
But I agree, I could easily display a "server does not support this feature..." message in case it cannot connect.
I would say we don't need a special endpoint for this, the wizard tries to receive a new a new secret hash, if it receives a 404 we know the code is not there yet.

HazAT commented:
@getsentry/design I've added a new state to the `LoadingIndicator` had to do some refactoring of the CSS but I checked and it looks the same. Also added it to the storybook/styleguide.
Please review just this and tell me if it's fine.

HazAT commented:
I moved the loading indicator into a seperate PR
https://github.com/getsentry/sentry/pull/6541

mattrobenolt commented on src/sentry/api/endpoints/setup_wizard.py:
```diff
@@ -0,0 +1,63 @@
+from rest_framework.response import Response
+
+from sentry import ratelimits
+from django.core.cache import cache
```
Since this is a permanent thing, let's not use this cache. This cache is ephemeral and won't error if it's down, so things will just silently fail.

Should use `sentry.cache:default_cache` instead, which is backed by redis.

mattrobenolt commented on src/sentry/api/endpoints/setup_wizard.py:
```diff
@@ -0,0 +1,63 @@
+            return Response(serialize(wizard_data))
+        else:
+            # This creates a new available hash url for the project wizard
+            rate_limited = ratelimits.is_limited(
```
What's the intention here? This is globally limiting to 10 times per minute. I think you'd want to scope it by IP or something. This is way too restrictive for a global limit.

HazAT replied:
Thanks, I've added users ip address to the rate limit

mattrobenolt commented on src/sentry/api/urls.py:
```diff
@@ -761,6 +763,21 @@
+
+    # Project Wizard
+    url(
+        r'^wizard$',
```
End this with a `/` like all other endpoints plz.
